vault/router.go

```diff
@@ -802,20 +812,102 @@ func (r *Router) LoginPath(ctx context.Context, path string) bool {
 	remain := strings.TrimPrefix(adjustedPath, mount)
 
 	// Check the loginPaths of this backend
-	loginPaths := re.loginPaths.Load().(*radix.Tree)
-	match, raw, ok := loginPaths.LongestPrefix(remain)
-	if !ok {
+	pe := re.loginPaths.Load().(*loginPathsEntry)
+	match, raw, ok := pe.paths.LongestPrefix(remain)
+	if !ok && len(pe.wildcardPaths) == 0 {
+		// no match found
 		return false
 	}
-	prefixMatch := raw.(bool)
 
-	// Handle the prefix match case
-	if prefixMatch {
-		return strings.HasPrefix(remain, match)
+	// Matching Priority
+	//  1. prefix
+	//  2. exact
+	//  3. wildcard
+	if ok {
+		prefixMatch := raw.(bool)
+		if prefixMatch {
+			// Handle the prefix match case
+			return strings.HasPrefix(remain, match)
+		} else if match == remain {
+			// Handle the exact match case
+			return true
+		}
 	}
 
-	// Handle the exact match case
-	return match == remain
+	// check Login Paths containing wildcards
+	reqPathParts := strings.Split(remain, "/")
+	for wcPath := range pe.wildcardPaths {
+		if wcPath == "" {
+			continue
+		}
+		isPrefix := false
+
+		currWCPath := wcPath
+		if currWCPath[len(currWCPath)-1] == '*' {
+			isPrefix = true
+			currWCPath = currWCPath[0 : len(currWCPath)-1]
+		}
+
+		splitCurrWCPath := strings.Split(currWCPath, "/")
+
+		if len(reqPathParts) < len(splitCurrWCPath) {
+			// check if the path coming in is shorter; if so it can't match
+			continue
+		}
+		if !isPrefix && len(splitCurrWCPath) != len(reqPathParts) {
+			// If it's not a prefix we expect the same number of segments
+			continue
+		}
+
+		segments := make([]string, 0, len(splitCurrWCPath))
+		for i, wcPathPart := range splitCurrWCPath {
+			switch {
+			case wcPathPart == "+":
+				segments = append(segments, reqPathParts[i])
+
+			case wcPathPart == reqPathParts[i]:
+				segments = append(segments, reqPathParts[i])
+
+			case isPrefix && i == len(splitCurrWCPath)-1 && strings.HasPrefix(reqPathParts[i], wcPathPart):
+				segments = append(segments, reqPathParts[i:]...)
+			}
+		}
+		result := strings.Join(segments, "/")
+		if isPrefix {
+			return strings.HasPrefix(remain, result)
```

Comment:
Since the third case on the switch above expands on the request path to append the rest of the prefix to segments, do we need to prefix check here or can we simply rely on `result == remain` below for the check?

Reply:
Good point, and in fact this if check was causing the bug I was seeing. It would always return if the current wildcard path is a prefix even if the incoming request did not match. This was flaky though since it depended on the order of the values being ranged over in the map. And since this map was a map[string]bool the order is random when ranging over them.

Reply:
It turns out we do still need to do a prefix check. But we should not return unless `HasPrefix` also evaluates to true. Otherwise we should let the for loop continue to check other possibilities and only return false once all wildcard paths have been evaluated.